File: survival-c/rendering.c
#include "rendering.h"
#include <stdio.h>
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include "error.h"

void renderObject(SDL_Renderer* renderer, Object* obj) {
    SDL_Rect rect = {obj->position.x, obj->position.y, obj->width, obj->height};
    SDL_SetRenderDrawColor(renderer, 255, 0, 0, 255);  // Red color
    SDL_RenderFillRect(renderer, &rect);
}

SDL_Rect spriteClips[ANIMATION_FRAMES];

SDL_Texture* loadSpritesheet(SDL_Renderer* renderer, const char* path) {
    SDL_Surface* loadedSurface = IMG_Load(path);
    if (loadedSurface == NULL) {
      error(IMG_GetError());
    }

    SDL_Texture* spritesheet = SDL_CreateTextureFromSurface(renderer, loadedSurface);
    if (spritesheet == NULL) {
      error(SDL_GetError());
    }

    SDL_FreeSurface(loadedSurface);

    for (int i = 0; i < ANIMATION_FRAMES; i++) {
        int row = i / FRAME_COLS;
        int col = i % FRAME_COLS;
        
        spriteClips[i].x = col * SPRITE_WIDTH;
        spriteClips[i].y = row * SPRITE_HEIGHT;
        spriteClips[i].w = SPRITE_WIDTH;
        spriteClips[i].h = SPRITE_HEIGHT;
    }

    return spritesheet;
}

void renderCharacter(SDL_Renderer* renderer, SDL_Texture* spritesheet, int x, int y) {
    SDL_Rect renderQuad = { x, y, SPRITE_WIDTH, SPRITE_HEIGHT };
    frames_t currentFrame = getCurrentFrame();
    SDL_RenderCopy(renderer, spritesheet, &spriteClips[currentFrame], &renderQuad);
}

void renderFloor(SDL_Renderer* renderer, const char* path)
{
    int tileWidth = 45;
    int tileHeight = 20;

    SDL_Surface* surface = IMG_Load(path);
    if (!surface) {
        SDL_Log("Failed to load image: %s", IMG_GetError());
        return;
    }

    SDL_Texture* texture = SDL_CreateTextureFromSurface(renderer, surface);
    SDL_FreeSurface(surface);

    if (!texture) {
        SDL_Log("Failed to create texture: %s", SDL_GetError());
        return;
    }

    SDL_Rect srcRect = {
        .x = 16,
        .y = 200,
        .w = 45,
        .h = 20,
    };

    int tilesX = (WINDOW_WIDTH + tileWidth - 1) / tileWidth;
    int tilesY = (WINDOW_HEIGHT + tileHeight - 1) / tileHeight;

    for (int y = 0; y < tilesY; y++) {
        for (int x = 0; x < tilesX; x++) {
            SDL_Rect dstRect = {
                .x = x * tileWidth,
                .y = y * tileHeight,
                .w = tileWidth,
                .h = tileHeight
            };

            SDL_RenderCopy(renderer, texture, &srcRect, &dstRect);
        }
    }

    SDL_DestroyTexture(texture);
}

File: survival/src/Rendering.hs
module Rendering where

import qualified SDL
import qualified SDL.Image
import qualified SDL.Font as TTF
import qualified SDL.Video.Renderer as Renderer
import Control.Monad (forM_, when)
import Control.Exception (bracket)
import Foreign.C.String (withCString)
import Foreign.C.Types (CInt)
import Data.Vector (Vector)
import qualified Data.Vector as V
import Game


-- since were receiving everything as args no need to a loadEnemySpritesheet fn as in C
loadSpritesheet :: SDL.Renderer -> FilePath -> Int -> Int -> Int -> Int -> IO (SDL.Texture, Vector (SDL.Rectangle CInt))
loadSpritesheet renderer path animationFrames frameCols spriteWidth spriteHeight = do
    surface <- SDL.Image.load path
    texture <- SDL.createTextureFromSurface renderer surface
    SDL.freeSurface surface

    let spriteClips = V.generate animationFrames $ \i -> 
            let row = i `div` frameCols
                col = i `mod` frameCols
            in SDL.Rectangle
                (SDL.P (SDL.V2 (fromIntegral $ col * spriteWidth)
                           (fromIntegral $ row * spriteHeight)))
                (SDL.V2 (fromIntegral spriteWidth) (fromIntegral spriteHeight))

    return (texture, spriteClips)

renderCharacter :: SDL.Renderer 
                -> (SDL.Texture, Vector (SDL.Rectangle CInt))  -- This is what loadSpritesheet returns
                -> Int  -- x position
                -> Int  -- y position
                -> Int  -- Current frame
                -> Int  -- sprite width
                -> Int  -- sprite height
                -> IO ()
renderCharacter renderer (spritesheet, spriteClips) x y currentFrame spriteWidth spriteHeight = do
    let renderQuad = SDL.Rectangle 
            (SDL.P (SDL.V2 (fromIntegral x) (fromIntegral y)))
            (SDL.V2 (fromIntegral spriteWidth) (fromIntegral spriteHeight))
    
    let sourceRect = spriteClips V.! currentFrame
    SDL.copy renderer spritesheet (Just sourceRect) (Just renderQuad)


renderEnemy :: SDL.Renderer -> (SDL.Texture, Vector (SDL.Rectangle CInt)) 
               -> Enemy -> Int -> Int -> IO ()
renderEnemy renderer (spritesheet, enemyClips) enemy enemyWidth enemyHeight = do
  let renderQuad = SDL.Rectangle
          (SDL.P (SDL.V2 ( round $ x (ePosition enemy)) (round $ y (ePosition enemy))))
          (SDL.V2 (fromIntegral enemyWidth) (fromIntegral enemyHeight))
  
  let sourceRect = enemyClips V.! (fromEnum (currentFrame enemy))
  SDL.copy renderer spritesheet (Just sourceRect) (Just renderQuad)


-- render all enemies if enemy is not dead. pass enemyWidth = 36 and enemyHeight = 36 to renderEnemy
renderEnemies :: SDL.Renderer -> (SDL.Texture, Vector (SDL.Rectangle CInt)) -> [Enemy] -> IO ()
renderEnemies renderer enemySpritesheet enemies = do
  forM_ enemies $ \enemy ->
    when (not $ eDead enemy) $
      renderEnemy renderer enemySpritesheet enemy 36 36

renderFloor :: SDL.Renderer -> FilePath -> IO ()
renderFloor renderer path = do
  let tileWidth = 45
      tileHeight = 20
      windowWidth = 950
      windowHeight = 800

  surface <- SDL.Image.load path
  texture <- SDL.createTextureFromSurface renderer surface
  SDL.freeSurface surface

  let srcRect = SDL.Rectangle (SDL.P (SDL.V2 16 200)) (SDL.V2 45 20)  
      windowSize = SDL.V2 windowWidth windowHeight
      tilesX = (windowWidth + tileWidth - 1) `div` tileWidth
      tilesY = (windowHeight + tileHeight - 1) `div` tileHeight

  forM_ [0..tilesY-1] $ \y ->
    forM_ [0..tilesX-1] $ \x -> do
      let dstRect = SDL.Rectangle
              (SDL.P (SDL.V2 (fromIntegral (x * tileWidth)) (fromIntegral (y * tileHeight))))
              (SDL.V2 (fromIntegral tileWidth) (fromIntegral tileHeight))
      SDL.copy renderer texture (Just srcRect) (Just dstRect)
  SDL.destroyTexture texture










